### PR TITLE
Unload unused models from server memory

### PR DESCRIPTION
**Pull request by ricwo**
**Proposed changes**:
- only ever keep one model per project in the server's memory under `Project._models`
- unload previous NLU model from `Project` when a new model is requested by the server

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog

### CODE REVIEW

**Comment by wrathagom** [comment]
I haven't reviewed the code, but as I read the description this would be a big change for anyone using multiple fixed model names under the same project, which seems to be a common practice.

**Comment by ricwo** [comment]
This only affects the server's in-memory models. The change means the server's `project_store` only contains the latest-used model, otherwise if you keep switching models you might run out of memory. 

You can of course store as many models as you like, but when you switch them (i.e. `parse` requests a different model than the previous `parse`) they are newly loaded into memory 🙂

**Comment by wrathagom** [comment]
In our use case, and several others I am aware of, we use projects as individual bots and serve multiple models simultaneously from within the project. For example, we may have a pizza bot project and then a model inside for each language English, Spanish, German, etc. We manage our own versions with a version number suffixed to the model name.

Just trying to make you aware that what I am describing is a pretty common use case that would have to be re-designed by the implementation of your proposed method.

@tmbo @amn41 any input.

**Comment by amn41** [comment]
thanks for the feedback @wrathagom ! and great that you keep an eye on the issues before we break anything :)

the case we're trying to guard against is that if you are continuously running a server and training models, the old ones never get unloaded and Rasa's memory footprint will monotonically increase. 

I'm sure there's some logic we can implement that supports both use cases, possibly adding an `auto_onload` flag to the server config, or coming up with another scheme for loading / unloading

**Comment by ricwo** [comment]
I have made some changes that should also accommodate your use case, @wrathagom:
- fixed model names (those not beginning with `model_`) are unloaded iff their name differs from a newer one by a version number (the model name has to end on `_X` or `_vX` )
- default model names (`model_X`) are always unloaded

**Comment by wrathagom** [comment]
@ricwo @amn41 I understand the need for it, indeed it's been a community request for awhile. Just pointing out that there was an issue open related to this #438 , specifically having to do with exposing endpoints for managing the loading/deleting of models. If I had my way that would still be the solution, but then again we're building platforms on top of Rasa not just individual agents. Maybe my team can take the code from this PR and expand it a bit to add the endpoints and contribute that back.

I spoke a bit to soon on our specific use case as ours is actually a hash, which includes among other things the date. `Small Talk_294b4d80-5295-0b20-3b79-d30f686fb700`

**Comment by tmbo** [comment]
I think it makes sense to do the following:
- addd a `DELETE` endpoint to manually unload models (no deletion on disk, just unloading from mem) - this should be simple as we already have the code for the unloading (@ricwo let me know if there are issues)
- list the loaded models in the status endpoint (right now we list all, so it is not clear which ones are loaded - whatever is easiest here (e.g. adding an attribute, creating a separate property next to `available_projects`))
- don't do any automatic unloading

File: rasa_nlu/data_router.py [hunk 1]
```diff
@@ -346,3 +346,20 @@ def evaluate(self, data, project=None, model=None):
                 "f1_score": f1,
                 "accuracy": accuracy}
         }
+
+    def unload_model(self, project, model):
+        # type: (Text, Text) -> Dict[Text]
+        """Unload a model from server memory."""
+
+        project = project or RasaNLUConfig.DEFAULT_PROJECT_NAME
```
Comment (tmbo) [inline rasa_nlu/data_router.py]: should we use a default here?

File: rasa_nlu/data_router.py [hunk 2]
```diff
@@ -346,3 +346,20 @@ def evaluate(self, data, project=None, model=None):
...
+
+        if project not in self.project_store:
+            raise InvalidProjectError("Project {} could not "
+                                      "be found".format(project))
+
+        try:
+            unloaded_model = self.project_store[project].unload(model)
+            return {"unloaded_model": unloaded_model}
```
Comment (tmbo) [inline rasa_nlu/data_router.py]: I think it is fine to just return the model without the `"unloaded_model"` wrapper, so just `return unloaded_model`

File: rasa_nlu/project.py [hunk 3]
```diff
@@ -182,7 +185,13 @@ def _read_model_metadata(self, model_name):
 
     def as_dict(self):
         return {'status': 'training' if self.status else 'ready',
-                'available_models': list(self._models.keys())}
+                'available_models': list(self._models.keys()),
+                'loaded_models': self._list_loaded_models()}
+
+    def _list_loaded_models(self):
+        return [
+            model for model, interpreter in self._models.items() if interpreter
```
Comment (tmbo) [inline rasa_nlu/project.py]: `if interpreter is not None`

File: rasa_nlu/data_router.py (hunk 1 above)
Comment (tmbo) [reply rasa_nlu/data_router.py]: lets move this default to the endpoint level

File: rasa_nlu/server.py [hunk 4]
```diff
@@ -234,6 +233,25 @@ def evaluate(self, request):
...
+    def unload_model(self, request):
+        params = {
+            key.decode('utf-8', 'strict'): value[0].decode('utf-8', 'strict')
+            for key, value in request.args.items()
+        }
+
+        request.setHeader('Content-Type', 'application/json')
+        try:
+            request.setResponseCode(200)
+            response = self.data_router.unload_model(params.get('project'),
```
Comment (tmbo) [inline rasa_nlu/server.py]: `params.get("project", RasaNLUConfig.DEFAULT_PROJECT_NAME)`

File: rasa_nlu/server.py [hunk 5]
```diff
@@ -234,6 +233,25 @@ def evaluate(self, request):
...
+                                                     params.get('model'))
+            return simplejson.dumps(response)
+        except Exception as e:
+            request.setResponseCode(500)
```
Comment (tmbo) [inline rasa_nlu/server.py]: please also log to logger

**Pull request merged**